runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/CoderTranslator.java

```diff
@@ -0,0 +1,44 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.runners.core.construction;
+
+import java.util.List;
+import org.apache.beam.sdk.annotations.Experimental;
+import org.apache.beam.sdk.annotations.Experimental.Kind;
+import org.apache.beam.sdk.coders.Coder;
+
+/**
+ * An interface that translates coders to components and back.
+ *
+ * <p>This interface is highly experimental, and incomplete. Coders must in the general case have
+ * the capability to encode an additional payload, which is not currently supported. This exists as
+ * a temporary measure.
+ */
+@Experimental(Kind.CORE_RUNNERS_ONLY)
+public interface CoderTranslator<T extends Coder<?>> {
```

Comment:
The core SDK (and similar for other SDKs) can already shade proto - it is isolated to runners-core-construction for just this reason, where proto is always on the surface without major changes.

You have a good point for IO connectors and other libraries of transforms that need to register coder translators but cannot have proto on their API surface.

It is a very poor API, but mostly harmless, for the translator to provide:

 - Component coders (which will be recursively translated)
 - `SdkFunctionSpec.FunctionSpec.urn` as a string
 - `SdkFunctionSpec.FunctionSpec.parameter` as bytes
 - (and the environment for the `SdkFunctionSpec` is Java SDK automatically)

I'm annoyed that `FunctionSpec.parameter` in the Runner API protos is `Any` and it turns out that buys us nothing in this case (it may be used in a non-vacuous way in other contexts, but this issue is evidence that they shouldn't be shared). So there's a choice here about whether we say that these bytes are required to be parseable as an `Any` so that it can be a straightforward representation or whether we double box them as a proto `Bytes` message.